cmake/clang-format.cmake

```diff
@@ -1,36 +1,83 @@
+    else()
+        set(${validator_result_var} TRUE PARENT_SCOPE)
+    endif()
+endfunction()
+
+function(print_clang_format_install_instructions)
+    message(STATUS "🛠️ Please install clang-format to enable code formatting")
+    if(UNIX)
+        if(APPLE)
+            message(STATUS "🍎 On MacOS, you can install clang-format with:")
+            message(STATUS "\t> brew install clang-format")
+        else()
+            message(STATUS "🐧 On Ubuntu, you can install clang-format with:")
+            message(STATUS "\t> sudo apt-get install clang-format")
+        endif()
+    elseif(WIN32)
+        message(STATUS "🪟 On Windows, you can install clang-format with:")
+        message(STATUS "\t> winget llvm")
+    endif()
+endfunction()
+
+find_program(CLANG_FORMAT clang-format
+             VALIDATOR check_clang-format_version)
 
 if(NOT CLANG_FORMAT)
     message(WARNING "❗clang-format not found")
-    return()
-endif()
-
-# list all files to format
-set(
-    FORMAT_PATTERNS
-    include/*.hpp
-    test/*.cpp
-    test/*.hpp
-    CACHE STRING
-    "; separated patterns relative to the project source dir to format"
-)
-
-set(ALL_FILES_TO_FORMAT "")
-foreach(PATTERN ${FORMAT_PATTERNS})
-    file(GLOB_RECURSE FILES_TO_FORMAT ${CMAKE_SOURCE_DIR}/${PATTERN})
-    list(APPEND ALL_FILES_TO_FORMAT ${FILES_TO_FORMAT})
-endforeach()
-
-add_custom_target(
-    clang-format
-    COMMAND ${CLANG_FORMAT} -i -style=file ${ALL_FILES_TO_FORMAT}
-    WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
-    COMMENT "Running clang-format on all files"
-)
-
-add_custom_target(
-    clang-format_dry_run
-    COMMAND ${CLANG_FORMAT} --dry-run -style=file ${ALL_FILES_TO_FORMAT}
-    WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
-    COMMENT "Running dry clang-format on all files"
-)
+
+    print_clang_format_install_instructions()
+else()
+    get_clang_format_version(${CLANG_FORMAT})
+    message(STATUS "✅clang-format found at ${CLANG_FORMAT} | version: ${CLANG_FORMAT_MAJOR_VERSION}.${CLANG_FORMAT_MINOR_VERSION}.${CLANG_FORMAT_PATCH_VERSION}")
+
+    # list all files to format
+    set(
+        FORMAT_PATTERNS
+        include/*.hpp
+        test/*.cpp
+        test/*.hpp
+        CACHE STRING
+        "; separated patterns relative to the project source dir to format"
+    )
+
+    set(ALL_FILES_TO_FORMAT "")
+    foreach(PATTERN ${FORMAT_PATTERNS})
+        file(GLOB_RECURSE FILES_TO_FORMAT ${CMAKE_SOURCE_DIR}/${PATTERN})
+        list(APPEND ALL_FILES_TO_FORMAT ${FILES_TO_FORMAT})
+    endforeach()
+
+    add_custom_target(
+        clang-format
+        COMMAND ${CLANG_FORMAT} -i -style=file ${ALL_FILES_TO_FORMAT}
+        WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
+        COMMENT "Running clang-format on all files"
+    )
+
+    add_custom_target(
+        clang-format_dry_run
+        COMMAND ${CLANG_FORMAT} --dry-run -style=file ${ALL_FILES_TO_FORMAT}
+        WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
+        COMMENT "Running dry clang-format on all files"
+    )
```

Comment:
What is the purposes of running a dry run after an actual run?

Is it to check that some files cannot be formatted?

Reply:
A dry run will just report the place where the code is not well formated.
As de developer, the workflow is:
Run the dry run and possibly add some no-lint in the code to avoid auto format.
Run clang format auto fix

Reply:
Actually you can forget about my comment since it is irrelevant: I thought those targets were run in this order as well which is not the case.